### DISCUSSION TITLE
Orbot keeps changing identity

### DISCUSSION THREAD

**Discussion by Pinkolik** [guardianproject/orbot-android#719] [Q&A]
Hello! I'm working on a feature for an app where all other apps' traffic is redirected through Orbot proxy. But Orbot keeps changing exit node (identity) for some reason. Is it normal behavior? Can I prevent Orbot from doing that? Here's some logs
![image](https://user-images.githubusercontent.com/26690061/181726220-9bd002a4-adfd-4f42-ac15-ab03f002577d.png)

**Comment by n8fr8** [comment]
Tor is meant to create new circuits (which includes exits) after a certain amount of time... maybe 10 minutes. That shouldn't really matter to your app though... it would just look like a fresh IP address to the remote server. Do you need to keep a stable IP address for your app?

Otherwise, the log output you are showing above seems like a different problem. Is there some issue with your server blocking connections from Tor?

**Reply by n8fr8** [reply]
cc'ing @bitmold into this chat!

**Reply by Pinkolik** [reply]
> Tor is meant to create new circuits (which includes exits) after a certain amount of time... maybe 10 minutes. That shouldn't really matter to your app though... it would just look like a fresh IP address to the remote server. Do you need to keep a stable IP address for your app?
> 
> Otherwise, the log output you are showing above seems like a different problem. Is there some issue with your server blocking connections from Tor?

I don't really need a stable IP. I just want Orbot to act as it intended. Actually you can check out my PR which uses Orbot. https://github.com/schwabe/ics-openvpn/pull/1512

It detects Orbot/Tor packets by destination IP address and matches it with known public nodes. Are there some other packets that should be allowed to pass in order to Orbot function properly?
 
I mean it's working and all but that log doesn't look healthy :)

**Reply by n8fr8** [reply]
Okay, so first you capture traffic in the OpenVPN client. If it is NOT Tor, then you proxy that into the Tor SOCKS port. If it IS Tor, then you route that over the VPN?

Hmm, I have to think about this some more. Interesting idea though!

**Reply by n8fr8** [reply]
Could you instead detect the UID of the packets as the come in, and match that to Orbot? I think there are too many potential remote servers that tor connects to for you to figure it out based on destination.

**Reply by Pinkolik** [reply]
> Could you instead detect the UID of the packets as the come in, and match that to Orbot? I think there are too many potential remote servers that tor connects to for you to figure it out based on destination.

That was my first implementation of this idea, I used [getConnectionOwnerUid](https://developer.android.com/reference/android/net/ConnectivityManager#getConnectionOwnerUid(int,%20java.net.InetSocketAddress,%20java.net.InetSocketAddress)) but it turned out to be a rather expensive operation to do on every packet, so I decided to use public nodes as a workaround to increase connection speed. By the way, even when I used to detect all Orbot traffic with getConnectionOwnerUid, I still had the same problem and same logs. Also, getConnectionOwnerUid only supports detecting TCP or UDP packets as well as my current implementation, could Orbot possibly use any other protocols which are not passed through?

**Reply by Pinkolik** [reply]
> Okay, so first you capture traffic in the OpenVPN client. If it is NOT Tor, then you proxy that into the Tor SOCKS port. If it IS Tor, then you route that over the VPN?

Yep, that's right

**Reply by Pinkolik** [reply]
@n8fr8, @bitmold, any ideas?

**Comment by Pinkolik** [comment]
Just bumping this thread :)
